docs/adr/0024-upgrade-templating-toolset-of-auth-server-emails.md

```diff
@@ -0,0 +1,261 @@
+- Pros
+  - The logicless nature of the templates will be eliminated once they're refactored in React
+  - We already use React in `fxa-settings`, `fxa-admin-panel`, `fxa-payments-server`, and `fxa-react`. There’s potential to cross-share components with the new email templates.
+  - React offers us many integration options. E.g there is one with [MJML](https://github.com/wix-incubator/mjml-react)
+  - React has already been audited by our security team.
+- Cons
+  - It could be a overkill to use React, since the library comes with far too many features which may not be needed for templating
+  - The templates in the `fxa-auth-server` do not contain heavy business logic so building them with a frontend-heavy framework like React might complicate things
+  - Since we will be converting React components into static HTML pages, another layer of complexity as well as new dependencies would be added to the project
+
+### Option C - Use EJS
+
+- Description
+  - Template the auth server mail components using [EJS](https://ejs.co/)
+- Pros
+  - The syntax and logic of EJS is similar to JavaScript, leading to a very small learning curve.
+  - EJS has a large community of users which will ensure faster help and support.
+  - It is a simple templating engine and far easier to use as its plain Javascript and since we want to keep our mail templates straightforward, using EJS could be a good choice.
+- Cons
+  - Adds a new dependency to the stack
+  - Lacks support for `block` functionality, which will make resuing components a bit difficult.
+  - Although we can implement `block` functionality with the help of third-party support, most of the projects are outdated and not maintained anymore.
+
+### Pros and Cons of using MJML as part of our templating stack
+
+As part of modernizing the templates, we aim to replace the framework in which our templates are currently built in so as to include more logic in them while development. Along with the framework, we considered using [MJML](https://mjml.io/) as our potential markup language, designed to reduce the pain of coding a responsive email.
+
+- Pros
+  - MJML is a component-based library having its own pre-built components as well as gives us the ability to define custom compononents which translates to responsive HTML after compilation. These components can be reused across the templates and modified by passing parameters e.g. [this social links component](https://documentation.mjml.io/#mj-social)
+  - MJML can help us create responsive emails using mobile-first approach and it controls that responsiveness automatically needing very less to no developer intervention.
+  - This library has been created by keeping in mind all the industry standards as well as the quirks of old email clients, so we will eventually be saving time on investigating cross-browser compatability of our templates
```

Comment:
In my experience writing HTML emails, time spent verifying and addressing bugs in email rendering across various apps was rather painful. There's some services that make it easier to see how various email apps will show it now, but its definitely a pretty huge Pro to have a large library of known and working components.

Reply:
Maybe this should be the first point under pros. When I was looking into MJML, its biggest claim to fame seemed to be the responsiveness aspect, of which it didn't seem like we'd "need" much of until our templates were redesigned. I haven't seen many tickets come in for an email style being broken in whatever client (also haven't been watching FxA issues come in the last 8 months though) so maybe I didn't give this pro enough weight.

@bbangert you mentioned MJML has multiple previewing options, do you have one you like to use or a suggestion to see various states an email can be built in? I liked the Storybook option as I know we can use it with React or EJS and we can easily see all of the different available states + show those to design and product, but if there's something else we can use I'm all ears. 😄

Reply:
> @bbangert you mentioned MJML has multiple previewing options, do you have one you like to use or a suggestion to see various states an email can be built in? I liked the Storybook option as I know we can use it with React or EJS and we can easily see all of the different available states + show those to design and product, but if there's something else we can use I'm all ears. 😄

So, I'm not really sure how many different available states HTML emails should have. They're easy enough to mess up in email clients, that having minimal changes to sub in values, or maybe hide/show a component as applicable should be the most changes we'd want. I'd love to hear from @lmorchard on what additional features were missing from handlebars that were pain points in emails.

The way I've used it once (I know Mailjet has their own other way), was to use nunjucks with MJML. The VS Code plugin for MJML expanded the MJML tags, and left in the nunjuck syntax. This generally worked fine for proofing in VS Code even with MJML component includes. There's also:
- https://mjmlio.github.io/mjml-app/
- https://mjml.io/try-it-live

Though obviously those might have more difficulty with MJML includes if the included component file paths are inaccessible, while the VS Code plugin running locally was able to preview the email correctly. Using EJS might require us to compile the template before it can be previewed, as the syntax will likely mess up MJML preview tools. That would be interesting to explore.

Reply:
> So, I'm not really sure how many different available states HTML emails should have. 

You mean "should" as in for best practice? Yeah, that's interesting. I wish I knew how many states our current emails could be in.

> Using EJS might require us to compile the template before it can be previewed, as the syntax will likely mess up MJML preview tools. That would be interesting to explore.

I kind of wonder if we used EJS + MJML if we could have some sort of "build all emails and states" command that we could run when starting Storybook up which would allow us to do something like `import Component from './build/index.html';` and then still use React in Storybook, but... I dunno, that also sounds messy.